cmd/generate-catalog/types.go

```diff
@@ -0,0 +1,279 @@
+// it will be represented in YAML like this:
+//   - schema: olm.channel
+//     name: rhacs-3.64
+//     package: rhacs-operator
+//     entries:
+//   - <ChannelEntry>
+func newChannel(version *semver.Version, entries []ChannelEntry) *Channel {
+	return &Channel{
+		Schema:  "olm.channel",
+		Name:    generateChannelName(version),
+		Package: "rhacs-operator",
+		Entries: entries,
+	}
+}
+
+// Create a special "olm.channel" object with name "latest".
+// It is a deprecated channel which was used before 4.x.x version.
+func newLatestChannel(entries []ChannelEntry) Channel {
+	return Channel{
+		Schema:  "olm.channel",
+		Name:    "latest",
+		Package: "rhacs-operator",
+		Entries: entries,
+	}
+}
+
+// Create a special "olm.channel" object with name "stable".
+// It is a default channel for all versions after 4.x.x.
+func newStableChannel(entries []ChannelEntry) Channel {
+	return Channel{
+		Schema:  "olm.channel",
+		Name:    "stable",
+		Package: "rhacs-operator",
+		Entries: entries,
+	}
+}
+
+// Create a new Chanel entry object which should be added to Channel entries list.
+// it will be represented in YAML like this:
+//   - name: rhacs-operator.v<version>
+//     replaces: rhacs-operator.v<previousEntryVersion>
+//     skipRange: '>= <previousChannelVersion> < <version>'
+//     skips:
+//   - rhacs-operator.v4.1.0
+func newChannelEntry(version, previousEntryVersion, previousChannelVersion *semver.Version, brokenVersions []*semver.Version) ChannelEntry {
+	entry := ChannelEntry{
+		Name: generateBundleName(version),
+	}
+	entry.addReplaces(version, previousEntryVersion)
+	entry.addSkipRange(version, previousChannelVersion)
+	entry.addSkips(version, brokenVersions)
+	return entry
+}
+
+func (e *ChannelEntry) addReplaces(version, previousEntryVersion *semver.Version) {
+	// skip setting "replaces" key for specific versions
+	versionsWithoutReplaces := []string{"4.0.0", "3.62.0"}
+
+	replacesVersion := previousEntryVersion.Original()
+	if !slices.Contains(versionsWithoutReplaces, version.Original()) {
+		e.Replaces = "rhacs-operator.v" + replacesVersion
+	}
+}
+
+func (e *ChannelEntry) addSkipRange(version, previousChannelVersion *semver.Version) {
+	skipRangeGreaterThanEqual := fmt.Sprintf("%d.%d.0", previousChannelVersion.Major(), previousChannelVersion.Minor())
+	skipRangeLessThan := version.Original()
+	e.SkipRange = fmt.Sprintf(">= %s < %s", skipRangeGreaterThanEqual, skipRangeLessThan)
+}
+
+func (e *ChannelEntry) addSkips(version *semver.Version, brokenVersions []*semver.Version) {
+	for _, brokenVersion := range brokenVersions {
+		// for any broken X.Y.Z version add "skips" for all versions > X.Y.Z and < X.Y+2.0
+		skipsUntilVersion := semver.MustParse(fmt.Sprintf("%d.%d.0", brokenVersion.Major(), brokenVersion.Minor()+2))
+		if version.GreaterThan(brokenVersion) && version.LessThan(skipsUntilVersion) {
+			e.Skips = append(e.Skips, fmt.Sprintf("rhacs-operator.v%s", brokenVersion.Original()))
```

Comment:
This can be changed to do:
```suggestion
			e.Skips = append(e.Skips, generateBundleName(brokenVersion))
```
after my other suggestion.